staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature_support_checker.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/version"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/component-base/featuregate"
+)
+
+var (
+	// This is a error type that we used in isDeprecated
+	ErrDeprecatedCluster = errors.New("")
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+)
+
+// FeatureSupportChecker to define Supports function which checks etcd with assigned feature and return boolean.
+type FeatureSupportChecker interface {
+	Supports(ctx context.Context, feature featuregate.Feature) (bool, error)
+}
+
+type DefaultFeatureSupportChecker struct {
+	client *clientv3.Client
+}
+
+func newDefaultFeatureSupportChecker(client *clientv3.Client) *DefaultFeatureSupportChecker {
+	return &DefaultFeatureSupportChecker{
+		client: client,
+	}
+}
+
+// Supports checks whether a specific feature is supported by etcd.
+// TODO: Extend this implementation to support feature-checking for the cacher.
+func (c *DefaultFeatureSupportChecker) Supports(ctx context.Context, feature featuregate.Feature) (bool, error) {
+	if feature == genericfeatures.ConsistentListFromCache {
+		return isDeprecated(ctx, c.client)
```

Comment:
Information what features etcd supports will be used in every LIST request, calling etcd Status should not be super costly, but also it's not free and we should limit those requests. Please also remember that apiserver creates a storage for every resource which there might be tens. 

Approaches I see: 
* We leave catching up to the client, which for now should be just watch cache. We might cache it there, however this leaves a gap for future where a new place starts calling Status for different reason and forgets.
* We cache the information on storage, so during bootstrap this can result in a lot of requests sent at once.
* We could have a global featureSupportChecker that caches the value per each endpoint.

Reply:
I definitely won't this to happen periodically (at least for now). I discussed that with @deads2k some time ago and agreed that checking that on startup is acceptable compromise here.

At the same time, the cost of that request is low enough that I can accept the cost of a call per storage.

So I would really prefer to simply call it once (with retries in case of failures) and use the same value later.

Reply:
Also i put some caching in the Supports to reference this issue, please take a look at it it might help.